Adding artist screen features

tolmadima opened the pull request.

tolmadima committed:
Added dialog for user on request Error

tolmadima committed:
On item click showing clicked artist

tolmadima committed:
Update Artist.java

ashvidchenko commented on app/src/main/java/com/example/lastfm/Artist.java:
```diff
@@ -6,11 +6,12 @@
 
-
-    public Artist(String name, String playCount) {
+    public Artist(String name, String playCount,String imageUrl) {
```
u should use different POJO classes for different requests

tolmadima replied:
I use this pojo only for mainactivity, i was thinking how to add actual pictures to recycler view. Deleted.

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistScreen.java:
```diff
@@ -2,13 +2,29 @@
 
+    TextView tvNameView;
+    TextView tvPlayCount;
+    ImageView artistImage;
```
all these variables must be private

tolmadima replied:
Set to private

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistScreen.java:
```diff
@@ -2,13 +2,29 @@
+        Intent intent = getIntent();
+
+        String name = intent.getStringExtra("artistName");
+        String playCount = intent.getStringExtra("artistPlayCount");
```
why do u need this info from previous activity?

tolmadima replied:
I need only name to make request

tolmadima replied:
PlayCount deleted

ashvidchenko commented on app/src/main/java/com/example/lastfm/MainActivity.java:
```diff
@@ -66,6 +69,8 @@ public void onSuccess(List<Artist> value) {
 
                         @Override
                         public void onError(Throwable e) {
+                            Context context = getApplicationContext();
```
use a context of the activity

tolmadima replied:
Changed.

ashvidchenko replied:
Never save context to a variable, get the context only at the moment when it needs

tolmadima replied:
deleted context variable.

ashvidchenko commented on app/src/main/java/com/example/lastfm/MainActivity.java:
```diff
@@ -66,6 +69,8 @@ public void onSuccess(List<Artist> value) {
                         @Override
                         public void onError(Throwable e) {
+                            Context context = getApplicationContext();
+                            Toast.makeText(context, "Ошибка получения списка артистов", Toast.LENGTH_LONG).show();
```
u must use a constant instead of raw string

tolmadima replied:
Changed.

ashvidchenko replied:
The name "toastError" says nothing about the type of error and the error message must be located in strings file from resource dir.

tolmadima replied:
renamed to requestErrorText

ashvidchenko commented on app/src/main/java/com/example/lastfm/MainActivity.java:
```diff
@@ -85,6 +90,10 @@ private void initRecyclerView() {
     @Override
     public void onArtistClick(int position) {
         Intent intent = new Intent(this,ArtistScreen.class);
+        Context context = getApplicationContext();
```
use a context of the activity

tolmadima replied:
Created activity context and use it now.

ashvidchenko replied:
Nothing is changed. But the solution described by u is wrong.

tolmadima replied:
Changed.

ashvidchenko commented on app/src/main/java/com/example/lastfm/MainActivity.java:
```diff
@@ -66,6 +69,8 @@ public void onSuccess(List<Artist> value) {
                         public void onError(Throwable e) {
+                            Context context = getApplicationContext();
+                            Toast.makeText(context, "Ошибка получения списка артистов", Toast.LENGTH_LONG).show();
                             e.printStackTrace();
```
remove logs from result code

tolmadima replied:
Removed.

ashvidchenko commented on app/src/main/java/com/example/lastfm/MainActivity.java:
```diff
@@ -85,6 +90,10 @@ private void initRecyclerView() {
+        Context context = getApplicationContext();
+        Toast.makeText(context,"position = " + position, Toast.LENGTH_LONG).show();
+        intent.putExtra("artistName",requestedArtists.get(position).getArtistName());
+        intent.putExtra("artistPlayCount",requestedArtists.get(position).getPlayCount());
```
u should call this code only once (requestedArtists.get(position))

tolmadima replied:
There is no second putExtra now.

ashvidchenko commented on app/src/main/java/com/example/lastfm/MainActivity.java:
```diff
@@ -85,6 +90,10 @@ private void initRecyclerView() {
     public void onArtistClick(int position) {
         Intent intent = new Intent(this,ArtistScreen.class);
+        Context context = getApplicationContext();
+        Toast.makeText(context,"position = " + position, Toast.LENGTH_LONG).show();
```
it's logging info for debugging. Remove it

tolmadima replied:
Removed.

tolmadima committed:
Added second method

ashvidchenko commented on app/src/main/java/com/example/lastfm/LastFMClient.java:
```diff
@@ -12,6 +12,8 @@
     @GET("?method=chart.gettopartists")
     Single<List<Artist>> getArtists(@Query("limit")int numberArtists, @Query("api_key")String key, @Query("format")String format);
+    @GET("?method=artist.getinfo")
+    Single<List> getArtistInfo(@Query("artist")String name,@Query("api_key")String key,@Query("format")String format);
```
why do u get only the list without any type?

tolmadima replied:
Changed List to ArtistInfo.

tolmadima committed:
Added Pojo

tolmadima committed:
Trying to make Pojo work

tolmadima committed:
Stuck

2020-04-01 16:39:46.926 25006-25006/com.example.lastfm W/System.err: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String com.example.lastfm.Artist.getArtistName()' on a null object reference
2020-04-01 16:39:46.927 25006-25006/com.example.lastfm W/System.err:     at com.example.lastfm.ArtistScreen$1.onSuccess(ArtistScreen.java:49)
2020-04-01 16:39:46.927 25006-25006/com.example.lastfm W/System.err:     at com.example.lastfm.ArtistScreen$1.onSuccess(ArtistScreen.java:40)
2020-04-01 16:39:46.927 25006-25006/com.example.lastfm W/System.err:     at io.reactivex.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run(SingleObserveOn.java:81)
2020-04-01 16:39:46.927 25006-25006/com.example.lastfm W/System.err:     at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:111)

tolmadima committed:
Current version

tolmadima committed:
Created POJO with plugin

tolmadima committed:
Artist Screen info about clicked artist added

tolmadima committed:
Fix

tolmadima committed:
Fixed cosmetic changes of items on ArtistScreen

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistInfo.java:
```diff
@@ -0,0 +1,416 @@
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+
+class ArtistData {
```
split all these classes on different files

tolmadima replied:
Splited.

ashvidchenko replied:
unit all these classes in one distinct package

tolmadima replied:
Created package and moved all these classes

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistScreen.java:
```diff
@@ -2,13 +2,77 @@
+        tvArtistBio = findViewById(R.id.big_artist_bio);
+        Intent intent = getIntent();
+        String name = intent.getStringExtra("artistName");
+        LastFMClient lastFMClient = ServiceGenerator.createService(LastFMClient.class);
```
read about singleton (there is a link in trello, in the card "read simple patterns") and fix it

tolmadima replied:
Added singleton

ashvidchenko replied:
Your implementation is wrong. Let's discuss it on the call today

tolmadima replied:
Changed implimentation

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistScreen.java:
```diff
@@ -2,13 +2,77 @@
+        Intent intent = getIntent();
+        String name = intent.getStringExtra("artistName");
+        LastFMClient lastFMClient = ServiceGenerator.createService(LastFMClient.class);
+        Single<ArtistInfo> call = lastFMClient.getArtistInfo(name, "b4ab3bf82dcb495e182e04cfc1f12b7b", REQUEST_TYPE).subscribeOn(Schedulers.computation()).observeOn(AndroidSchedulers.mainThread());
```
split this line on several lines

tolmadima replied:
Splited

ashvidchenko replied:
The first executing method must be located with the object on the same line

tolmadima replied:
Moved method.

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistScreen.java:
```diff
@@ -2,13 +2,77 @@
+                tvArtistBio.setText(bio);
+                tvPlayCount.setText(playcount);
+                Picasso.get().load(imageUrl).into(artistImage);
+                artistImage.setVisibility(imageUrl != null ? View.VISIBLE : View.GONE);
```
create a separate method for this callback

tolmadima replied:
Created separated method

ashvidchenko replied:
Re-read my comment and do what I asked u

tolmadima replied:
moved whole callback code to the method.

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistScreen.java:
```diff
@@ -2,13 +2,77 @@
+                tvArtistBio.setText(bio);
+                tvPlayCount.setText(playcount);
+                Picasso.get().load(imageUrl).into(artistImage);
+                artistImage.setVisibility(imageUrl != null ? View.VISIBLE : View.GONE);
```
create a separate method for this callback

tolmadima replied:
Created method loadArtist

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistScreen.java:
```diff
@@ -2,13 +2,77 @@
+            @Override
+            public void onError(Throwable e) {
+                e.printStackTrace();
+                Context context = getApplicationContext();
```
use activity context

tolmadima replied:
Changed.

ashvidchenko replied:
Your implementation is wrong

tolmadima replied:
Changed to *activityname*.this in both cases.

ashvidchenko commented on app/src/main/java/com/example/lastfm/LastFMClient.java:
```diff
@@ -12,6 +12,8 @@
     @GET("?method=chart.gettopartists")
     Single<List<Artist>> getArtists(@Query("limit")int numberArtists, @Query("api_key")String key, @Query("format")String format);
+    @GET("?method=artist.getinfo")
+    Single<ArtistInfo> getArtistInfo(@Query("artist")String name,@Query("api_key")String key,@Query("format")String format);
```
split params on several lines

tolmadima replied:
Splited

tolmadima committed:
Fixes

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistScreen.java:
```diff
@@ -2,13 +2,78 @@
+    TextView tvNameView = findViewById(R.id.big_artist_name);
+    TextView tvPlayCount = findViewById(R.id.big_artist_playcount);
+    ImageView artistImage = findViewById(R.id.big_artist_image);
+    TextView tvArtistBio = findViewById(R.id.big_artist_bio);
```
These variables must be private

tolmadima replied:
Setted to private

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistScreen.java:
```diff
@@ -2,13 +2,78 @@
+        LastFMClient lastFMClient = ServiceGenerator.createService(LastFMClient.class);
+        Single<ArtistInfo> call = lastFMClient
+                .getArtistInfo(name, "b4ab3bf82dcb495e182e04cfc1f12b7b", REQUEST_TYPE)
+                .subscribeOn(Schedulers.computation())
```
Network requests must be performed on Schedulers.io()

tolmadima replied:
Changed.

ashvidchenko commented on app/src/main/java/com/example/lastfm/MainActivity.java:
```diff
@@ -48,25 +51,26 @@ public void onClick(View view) {
-                    Single<List<Artist>> call = lastFMClient.getArtists(NUMBER_OF_ARTISTS, APP_ID, REQUEST_TYPE).subscribeOn(Schedulers.computation()).observeOn(AndroidSchedulers.mainThread());
+                    Single<List<Artist>> call = lastFMClient
+                            .getArtists(NUMBER_OF_ARTISTS, APP_ID, REQUEST_TYPE)
+                            .subscribeOn(Schedulers.computation())
```
Schedulers.io()

tolmadima replied:
Changed.

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistScreen.java:
```diff
@@ -2,13 +2,78 @@
+                Stats artistStat = info.getStats();
+                String playcount = artistStat.getPlaycount();
+                String bio = artistBio.getContent();
+                Image url = info.getImage().get(3);
```
Why do u get the third element?

tolmadima replied:
They counting from 1-4 (from small image to big) i needed medium size image.

ashvidchenko replied:
save this number as a constant

tolmadima replied:
created constant picSize.

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistScreen.java:
```diff
@@ -2,13 +2,78 @@
+        });
+    }
+
+    public void loadArtist(String name, String bio, String playcount, String imageUrl){
```
u don't load info so rename the method

tolmadima replied:
renamed to showArtist

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistScreen.java:
```diff
@@ -2,13 +2,78 @@
+                String bio = artistBio.getContent();
+                Image url = info.getImage().get(3);
+                String imageUrl = url.getText();
+                loadArtist(artistName,playcount,bio,imageUrl);
```
put the code of this callback to the separate method

tolmadima replied:
Moved all code to method showArtist

ashvidchenko commented on app/src/main/java/com/example/lastfm/LastFMClient.java:
```diff
@@ -11,7 +11,13 @@
+    @GET("?method=artist.getinfo")
+    Single<ArtistInfo> getArtistInfo(@Query("artist")String name,
+                                     @Query("api_key")String key,
+                                     @Query("format")String format);
```
remove params (api_key and format) from methods and create interceptor for adding them to requests. See the task "Interceptors" in Trello

tolmadima replied:
Added interceptor

tolmadima committed:
Comment fixes

ashvidchenko commented on app/src/main/java/com/example/lastfm/ArtistScreen.java:
```diff
@@ -2,13 +2,82 @@
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_artist_screen);
+         tvNameView = findViewById(R.id.big_artist_name);
```
use the same indent size everywhere

tolmadima replied:
Fixed

tolmadima committed:
removed .idea

tolmadima committed:
Trying to fix

tolmadima committed:
Error

tolmadima committed:
Trying to work with Interceptors

tolmadima committed:
Interceptor working

tolmadima committed:
Migrated Interceptor to another class

tolmadima committed:
Fix

tolmadima committed:
Fixed on click

tolmadima committed:
Changed names of items
